proposals/20210526-model_load_test_component.md

```diff
@@ -0,0 +1,43 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+**Your name:** Michal Brys
+
+**Your email:** michal.brys@openx.com
+
+**Your company/organization:** OpenX
+
+**Project name:** Model load test TFX Component
```

Comment:
@chongkong Thanks for the points, I thought about:
(1) Set up workers, send the requests and measure the latency distribution. The distribution metrics (min, max, avg, percentiles) could be published in the output artefact. It may allow us to compare models performance (if we'll keep the workers' number and traffic constant).
(2) From my experience, it's good to have a load test in a separate container to avoid competing for resources. But, it may not be required at this phase if we focus on the model and relative performance metrics.
(3) We may try to use the pubic Docker image for the load test and submit the custom values in params (i.e. https://ghz.sh/docs/usage). Custom Docker image is also an option with less dependency, but more maintaining effort.

Reply:
Sorry for the late follow-up! (Robert reminded me back to this thread)

1. So I assume it is not squeezing the model but test the moderate situation.. Or maybe we can frame it more generally as "any setting is okay; you can configure your model server and testing client setting to your taste, and this component simply provides bridge to that".
2. Agreed that model server should be on the separate container, and actually I was mainly asking about whether the "client" should be on the separate container. If client also runs on the separate container we would have total 3 containers running: 1 for component run, 1 for client, 1 for model server. If client runs on the same container that component runs, we would have total 2 containers: 1 for component + client run, 1 for model server. Unfortunately both options have a challenge which is...
3. (cont.) In either case, it is non-trivial how to bring load test binary (vegeta or ghz) to the container.
    1.  Case 1 (same container for client): Currently we uses `tensorflow/tfx` image for component container. Each component runs with the same image but with the different argument, so that entrypoint decides which component this container would run. If we use this same container as a load test client, such binary should be pre-installed in the container image. But I doubt we will do that for the standard `tensorflow/tfx` image because we cannot manage all dependencies for 3party addons. Instead, we should bake a new image (such as `tfx-addons/loadtest`) that is based on `tensorflow/tfx`, but with vegeta or ghz installed. But TFX team will probably be unable to maintain such container image.
    2.  Case 2 (separate container for client): If there is a separate container image with vegeta/ghz installed, we can use that container image for the load testing client. This is more affordable than case 1 because user can provide external container image as a config, and TFX team doesn't manage any extra container images. It would be even great if vegeta/ghz provides their own container images that TFX can use right away, but I don't know if they have ones. If we go with this option, I would like to ask you about more integration story included in this design.